rplugin/python3/deoplete/source/buffer.py

```diff
@@ -29,8 +29,12 @@ def on_event(self, context):
     def gather_candidates(self, context):
         self.__make_cache(context)
 
-        buffers = [x['candidates'] for x in self.__buffers.values()
-                   if x['filetype'] in context['filetypes']]
+        buffer_values = self.__buffers.values()
+        if context['vars'].get('require_same_filetype', True):
+          buffer_values = [x for x in buffer_values
```

Comment:
@sunaku This isn't entirely due to your changes, but this can be optimized like so:

```python
import itertools

...

        same_filetype = context['vars'].get('require_same_filetype', False)
        candidates = (x['candidates'] for x in self.__buffers.values()
                      if not same_filetype or
                      (same_filetype and x['filetype'] in context['filetypes']))
        return [{'word': x} for x in itertools.chain(*candidates)]
```

The old code + your changes has two list comprehensions (e.g. `[x for x in y]`).  With, the `functools.reduce()` line, you end up with 3 list iterations as well as `reduce()` creating new lists on each `add()`.

Using a generator comprehension (e.g. `(x for x in y)`), the results are evaluated when expanding `*candidates` on `itertools.chain()`.  In addition, the candidate lists are iterated and yielded to the final list comprehension (as opposed to concatenated *then* iterated).  Also, `functools.reduce()` throws an error on an empty list, `itertools.chain()` will just yield nothing, so the check for an empty list can be skipped.

I'm not generally uptight about optimizations like this, but I think it's worth shaving off a few nano/milliseconds for live completions.  You don't have add this code now if you don't want to.  But, I'll most likely add it myself in a separate PR if you don't 😜